File: P5018-noip-2018-T4.cpp
#include <iostream>
using namespace std;
struct node
{
    long long l, r, val;
} bt[1000002];
bool same(long long now1, long long now2)
{
    if (now1 == -1 && now2 == -1)
        return true;
    if (now2 == -1 || now1 == -1)
        return false;
    if (bt[now1].val != bt[now2].val)
        return false;
    return same(bt[now1].l, bt[now2].r) && same(bt[now1].r, bt[now2].l);
}
int count(long long now)
{
    if (now == -1)
        return 0;
    return count(bt[now].l) + count(bt[now].r) + 1;
}
int main()
{
    int n, ans = 0;
    cin >> n;
    for (int i = 1; i <= n; i++)
        cin >> bt[i].val;
    for (int i = 1; i <= n; i++)
        cin >> bt[i].l >> bt[i].r;
    for (int i = 1; i <= n; i++)
        if (same(i, i))
            ans = max(ans, count(i));
    cout << ans;
    return 0;
}

File: P5018-noip-2018-T4.py
def same(n1, n2) -> bool:
    if n1 is None and n2 is None:
        return True
    if n1 is None or n2 is None:
        return False
    if vals[n1] != vals[n2]:
        return False
    return same(lefts[n1], rights[n2]) and same(rights[n1], lefts[n2])


def count(n) -> int:
    if n is None:
        return 0
    return count(lefts[n]) + count(rights[n]) + 1


if __name__ == '__main__':
    n = int(input())
    vals = list(map(int, input().split()))
    lefts = []
    rights = []
    for i in range(n):
        l, r = map(int, input().split())
        if l == -1:
            l = None
        else:
            l = l - 1
        if r == -1:
            r = None
        else:
            r = r - 1
        lefts.append(l)
        rights.append(r)

    ans = 0
    for i in range(n):
        if same(i, i):
            ans = max(ans, count(i))
    print(ans)
